src/com/ms/silverking/cloud/meta/ExclusionSet.java

```diff
@@ -3,135 +3,69 @@
-                filtered.add(node);
-            }
-        }
-        return filtered;
-    }
+public class ExclusionSet extends ServerSetExtension {
+  private ExclusionSet(ServerSet serverSet, long mzxid) {
+    super(serverSet);
+    this.mzxid = mzxid;
+  }
 
-    public List<IPAndPort> filterByIP(Collection<IPAndPort> raw) {
-        List<IPAndPort>    filtered;
-        
-        filtered = new ArrayList<>(raw.size());
-        for (IPAndPort node : raw) {
-            boolean excluded;
-            
-            excluded = false;
-            for (String server : getServers()) {
-                if (node.getIPAsString().equals(server)) {
-                    excluded = true;
-                    break;
-                }
-            }
-            if (!excluded) {
-                filtered.add(node);
-            }
-        }
-        return filtered;
-    }
-    
-    public static ExclusionSet parse(String def) throws IOException {
-        return new ExclusionSet(new ServerSet(CollectionUtil.parseSet(def, singleLineDelimiter), VersionedDefinition.NO_VERSION));
-    }
-    
-    public static ExclusionSet parse(File file) throws IOException {
-        return new ExclusionSet(ServerSet.parse(new FileInputStream(file), VersionedDefinition.NO_VERSION));
-    }
+  public ExclusionSet(ServerSet serverSet) {
+    this(serverSet, INVALID_ZXID);
+  }
 
-    public static ExclusionSet union(ExclusionSet s1, ExclusionSet s2) {
-        ExclusionSet    u;
-        
-        u = emptyExclusionSet(0);
-        u = u.add(s1.getServers());
-        u = u.add(s2.getServers());
-        return u;
-    }
+  private ExclusionSet(long version) {
+    this(new ServerSet(new HashSet<>(), version));
+  }
+
+  public ExclusionSet(Set<String> excludedEntities, long version, long mzxid) {
+    this(new ServerSet(excludedEntities, version), mzxid);
+  }
+
+  public static ExclusionSet emptyExclusionSet(long version) {
+    return new ExclusionSet(version);
+  }
+
+  @Override
+  public ExclusionSet addByIPAndPort(Set<IPAndPort> newExcludedEntities) {
+    return (ExclusionSet) super.addByIPAndPort(newExcludedEntities);
+  }
+
+  @Override
+  public ExclusionSet add(Set<String> newExcludedEntities) {
+    return new ExclusionSet(serverSet.add(newExcludedEntities));
+  }
+
+  @Override
+  public ExclusionSet removeByIPAndPort(Set<IPAndPort> newExcludedEntities) {
+    return (ExclusionSet) super.removeByIPAndPort(newExcludedEntities);
+  }
+
+  @Override
+  public ExclusionSet remove(Set<String> newExcludedEntities) {
+    return new ExclusionSet(serverSet.remove(newExcludedEntities));
+  }
+
+  public static ExclusionSet parse(String def) {
+    return new ExclusionSet(
+        new ServerSet(CollectionUtil.parseSet(def, singleLineDelimiter), VersionedDefinition.NO_VERSION));
+  }
+
+  public static ExclusionSet parse(File file) throws IOException {
+    return new ExclusionSet(ServerSet.parse(new FileInputStream(file), VersionedDefinition.NO_VERSION));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when ServerSet.parse throws an exception. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).